### PR TITLE
Post-mortem April 11

### PR DESCRIPTION
**Pull request by saraycp**

### CODE REVIEW

**Review by danidoni — APPROVED** [review]

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+category: deployments
+---
+
+After implementing the feature that allows the users to comment on specific lines in the changes
+of a request, we also introduced the possibility to notify about them. However, there was a
+corner case which was not adapted and caused the notifications to fail, so our [reference server](https://build.opensuse.org) did not send any notifications during the Easter holidays.
+
+## Impact
+
+Our [reference server](https://build.opensuse.org) did not send any notifications (web or email) for a few days.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:13]: Both "Easter holidays" and "a few days" will leave a lot of people guessing which time frame this was. At least mention the dates please.

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+A user creating a comment on changes and subsequently the build service creating a notification for it. Code-wise, the introduction of comments on changes.
+
+## Detection
+
+Grafana alerted us with the information about the issue about 3 hours after it first happened, and then a few times in the following days.
+
+## Resolution
+
+A patch was introduced and then merged to handle the notifications for comments on changes properly.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:29]: Link to the PR?

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+Code didn't consider the new comments on changes, which made the delayed job crash so the application did not manage to send out emails and notifications.
+
+## Trigger
+
+A user creating a comment on changes and subsequently the build service creating a notification for it. Code-wise, the introduction of comments on changes.
+
+## Detection
+
+Grafana alerted us with the information about the issue about 3 hours after it first happened, and then a few times in the following days.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:25]: Link to obs-measure?

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+## Resolution
+
+A patch was introduced and then merged to handle the notifications for comments on changes properly.
+
+## Action items
+
+- Errors on the web notifications handling affected the email deliveries. It's important to split the code in the delayed job.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:33]: This is the first time you mention Emails *at all*. People should be aware about this without reading the fine print 😸

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+Errors do not always correlate with the latest deployed changes.
+
+## What Went Well?
+
+- Once the issue was analyzed thoroughly, we were able to come up with a patch that fully fixed the issue.
+- The patch was applied right before our users' activity was fully restarted after the Easter period.
+
+## What Went Wrong?
+
+- We usually correlate issues with the latest deployed changes. This time, there wasn't a direct match.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:47]: > We usually correlate issues with the latest deployed changes.

Huh? What? That is not really true, at least not for all of us I would say.

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+## Where We Got Lucky
+
+Not many people noticed due to the Easter holidays.
+
+## Timeline (times in UTC)
+
+- 06/04/2023 21:42 - A notification was created for the last time but its event wasn't marked as sent.
+- 07/04/2023 01:00 - First alert on Slack coming from Grafana.
+- 10/04/2023 09:21 - Start investigating the alerts and the causes (checking deployed changes, server updates, logs etc.)
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:63]: I would mention that this was the first regular working day, kind of to explain the huge gap between alert and investigation.

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: _posts/deployments/2023-04-12-post-mortem.md (hunk 4 above)
Comment (saraycp) [reply _posts/deployments/2023-04-12-post-mortem.md:33]: "Emails" was already mentioned in the `Impact` and `Root Causes` sections, but I'll make it clearer by mentioning it in the introduction.

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
+---
+layout: post
+title: "Post-mortem: Lack of Notifications from April 6 to April 11, 2023"
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:3]:
```suggestion
title: "Post-mortem: Lack of Emails/Notifications from April 6 to April 11, 2023"
```

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+category: deployments
+---
+
+After implementing the feature that allows the users to comment on specific lines in the changes
+of a request, we also introduced the possibility to notify about them. However, there was a
+corner case which was not adapted and caused the notifications to fail, so our [reference server](https://build.opensuse.org) did not send any web notifications or emails during the Easter holidays (from 6th to 11th April).
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:9]:
```suggestion
corner case which caused the notifications to fail, so our [reference server](https://build.opensuse.org) did not send any web notifications or emails during the Easter holidays (from 6th to 11th April).
```

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+## Impact
+
+Our [reference server](https://build.opensuse.org) did not send any notifications (web or email) for around 4 days and a half.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:13]:
```suggestion
Our [reference server](https://build.opensuse.org) did not send any notifications (web or email) for round about four and a half days.
```

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+## Root Causes
+
+Code didn't consider the new comments on changes, which made the delayed job crash so the application did not manage to send out emails and notifications.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:17]: If find this a bit too generic. Remember this is mostly for fellow rails practitioners (including our future selves). How about something like...

```suggestion
A `BsRequest` in OBS can have one or more `BSRequestAction`. Each `BsRequestAction` can generate it's own diff. So with the new [diff comments feature](/2023/03/07/request-workflow-redesign/) the thing we create `Notification` about is an individual `BsRequestAction` and not the generic `BsRequest`. 

The code we run to figure out who this `Notification` concerns (`NotifiedProjects`) did not consider this change, which made the `ActiveJob` that tries to generate the `Notification` crash. As the generation of `Notification` and the sending of Emails happens in the same job the application did not manage to send out emails or create notifications. 💥 
```

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+## Trigger
+
+A user creating a comment on changes and subsequently the build service creating a notification for it. Code-wise, the introduction of comments on changes.
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:21]:
```suggestion
Users creating comments on diffs.
```

File: _posts/deployments/2023-04-12-post-mortem.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+Not many people noticed due to the Easter holidays.
+
+## Timeline (times in UTC)
+
+- 06/04/2023 21:42 - A notification was created for the last time but its event wasn't marked as sent.
+- 07/04/2023 01:00 - First alert on Slack coming from Grafana, during Easter period.
+- 10/04/2023 09:21 - First regular working day after Easter. Start investigating the alerts.
+- 10/04/2023 10:30 - Check the email delivery is not broken by sending an event email from the Rails console.
+- 10/04/2023 10:52 on - Unsuccessful attempts to find the root cause and reproduce it in development environment (checking deployed changes, server updates, logs etc.)
```
Comment (hennevogel) [inline _posts/deployments/2023-04-12-post-mortem.md:65]: on?

**Review by hennevogel — CHANGES_REQUESTED** [review]

**Review by hennevogel — APPROVED** [review]

**Pull request merged**